File: 编程/leetcode/946-[M]-验证栈序列/solution.py
from typing import List


class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        stack = []
        p_pop = 0
        for num in pushed:
            stack.append(num)
            while stack and p_pop < len(popped) and stack[-1] == popped[p_pop]:
                stack.pop()
                p_pop += 1
        if p_pop == len(popped):
            return True
        return False



if __name__ == '__main__':
    test = Solution()
    print(test.validateStackSequences([1,2,3,4,5], [4,5,3,2,1]))
    print(test.validateStackSequences([1,2,3,4,5], [4,3,5,1,2]))
    print(test.validateStackSequences([1], [1]))
    print(test.validateStackSequences([1], [2]))
        

File: 编程/leetcode/946-[M]-验证栈序列/solution.rs
struct Solution;

impl Solution {
    pub fn validate_stack_sequences(pushed: Vec<i32>, popped: Vec<i32>) -> bool {
        let mut stack: Vec<i32>= Vec::new();
        let mut p_pop: usize = 0;
        for &num in pushed.iter() {
            stack.push(num);
            while stack.len() > 0 && stack[stack.len()-1] == popped[p_pop] {
                stack.pop();
                p_pop += 1;
            }
        }
        if stack.len() == 0 {
            return true;
        }
        return false;
    }
}


fn main() {
    println!("{}", Solution::validate_stack_sequences(vec![1,2,3,4,5], vec![4,5,3,2,1]));
    println!("{}", Solution::validate_stack_sequences(vec![1,2,3,4,5], vec![4,3,5,1,2]));
}